Calling .hasNext() or .next() on MultiPart Iterator after the first call to .next() causes NullPointerException

When iterating over the parts of a multipart request using the `io.helidon.http.media.multipart.MultiPart` iterator, calliing `.hasNext()` of `.next()` after the first call to  `.next()` causes a `NullPointerException` to be thrown.


## Environment Details
* Helidon Version: 4.0.5
* Helidon SE
* JDK version: 21
* OS: Linux x86
* Docker version (if applicable): /

----------

## Problem Description
**Current behavior:** Iterating over parts of a multipart requests fails when calling `.hasNext()` or `.next()` on `io.helidon.http.media.multipart.MultiPart` after `.next()` was called once.

**Expected behavior:** Calling `.hasNext()` should return `true` as long as unread parts remain and a subsequent call to `.next()` should return the next `io.helidon.http.media.multipart.ReadablePart` instance from the request. If no more parts remain, `.hasNext()` should return `false`.

Theis can be reproduced by running the [Helidon SE MultiPart Example](https://github.com/helidon-io/helidon/tree/main/examples/media/multipart) or the minimal example shown below.

When I did this, the following stacktrace was printed:
```
io.helidon.http.RequestException: Cannot invoke "io.helidon.common.buffers.BufferData.available()" because "this.nextBuffer" is null
	at io.helidon.http.RequestException$Builder.build(RequestException.java:139)
	at io.helidon.webserver.http.ErrorHandlers.unhandledError(ErrorHandlers.java:202)
	at io.helidon.webserver.http.ErrorHandlers.lambda$handleError$1(ErrorHandlers.java:182)
	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:198)
	at io.helidon.webserver.http.ErrorHandlers.handleError(ErrorHandlers.java:181)
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:118)
	at io.helidon.webserver.http.Filters.filter(Filters.java:77)
	at io.helidon.webserver.http.HttpRouting.route(HttpRouting.java:109)
	at io.helidon.webserver.http1.Http1Connection.route(Http1Connection.java:396)
	at io.helidon.webserver.http1.Http1Connection.handle(Http1Connection.java:169)
	at io.helidon.webserver.ConnectionHandler.run(ConnectionHandler.java:155)
	at io.helidon.common.task.InterruptableTask.call(InterruptableTask.java:47)
	at io.helidon.webserver.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:239)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:311)
Caused by: java.lang.NullPointerException: Cannot invoke "io.helidon.common.buffers.BufferData.available()" because "this.nextBuffer" is null
	at io.helidon.http.media.multipart.ReadablePartNoLength$PartInputStream.finish(ReadablePartNoLength.java:198)
	at io.helidon.http.media.multipart.ReadablePartNoLength.finish(ReadablePartNoLength.java:77)
	at io.helidon.http.media.multipart.MultiPartImpl.hasNext(MultiPartImpl.java:70)
	at io.helidon.examples.media.multipart.FileService.upload(FileService.java:136)
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.doRoute(HttpRouting.java:668)
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.call(HttpRouting.java:627)
	at io.helidon.webserver.http.HttpRouting$RoutingExecutor.call(HttpRouting.java:605)
	at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:75)
	... 8 more
``` 

## Steps to reproduce
[//]: # "Step by step instructions to reproduce the problem"
[//]: # "Provide sample code/application if relevant"

1. Create a new Project using this code snippet and run it
    ```java
    import io.helidon.http.media.multipart.MultiPart;
    import io.helidon.webserver.WebServer;
    
    public class Main {
        public static void main(String[] args) {
            WebServer.builder()
                    .port(8080)
                    .routing(routing -> routing.post("/", (serverRequest, serverResponse) -> {
                        var parts = serverRequest.content().as(MultiPart.class);
                        while (parts.hasNext()) {
                            var part = parts.next();
                            System.out.println(part.name());
                        }
                        serverResponse.status(200).send();
                    })).build().start();
            while (true) {}
        }
    }
    ```
2. Send a POST multipart request with more than 2 parts, e.g. using:
    ```bash
    curl --request POST 'http://localhost:8080' --form 'field_1=a' --form 'field_2=b'
    ```

The expected behavior of this procedure would be that the following output is printed:
```
field_1
field_2
```

The actual behavior is that only the first line including `field_1` is printed, then the NullPointerException is thrown.

--- a/http/media/multipart/src/main/java/io/helidon/http/media/multipart/ReadablePartLength.java
+++ b/http/media/multipart/src/main/java/io/helidon/http/media/multipart/ReadablePartLength.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2022, 2024 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -55,6 +55,21 @@ class ReadablePartLength extends ReadablePartAbstract {
     public <T> T as(GenericType<T> type) {
         return context.reader(type, partHeaders())
                 .read(type, inputStream(), partHeaders());
+    }
+
+    @Override
+    public void consume() {
+        if (inputStream == null) {
+            inputStream = inputStream();
+        }
+        try {
+            byte[] buffer = new byte[2048];
+            while (inputStream.read(buffer) > 0) {
+                // ignore
+            }
+        } finally {
+            inputStream.close();
+        }
     }
 
     @Override

--- a/http/media/multipart/src/main/java/io/helidon/http/media/multipart/ReadablePartNoLength.java
+++ b/http/media/multipart/src/main/java/io/helidon/http/media/multipart/ReadablePartNoLength.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2022, 2024 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -59,6 +59,21 @@ class ReadablePartNoLength extends ReadablePartAbstract {
     public <T> T as(GenericType<T> type) {
         return context.reader(type, partHeaders())
                 .read(type, inputStream(), partHeaders());
+    }
+
+    @Override
+    public void consume() {
+        if (inputStream == null) {
+            inputStream = inputStream();
+        }
+        try {
+            byte[] buffer = new byte[2048];
+            while (inputStream.read(buffer) > 0) {
+                // ignore
+            }
+        } finally {
+            inputStream.close();
+        }
     }
 
     @Override
@@ -167,6 +182,8 @@ class ReadablePartNoLength extends ReadablePartAbstract {
                     BufferData untilEol = dataReader.getBuffer(nextEol);
                     String theString = untilEol.readString(nextEol, StandardCharsets.US_ASCII);
                     if (boundary.equals(theString) || endBoundary.equals(theString)) {
+                        // we cannot skip the beginning of the boundary (as we call getBuffer, not readBuffer)
+                        trailingEol = false;
                         finished = true;
                         return null;
                     }
@@ -195,7 +212,10 @@ class ReadablePartNoLength extends ReadablePartAbstract {
         private void finish() {
             while (!finished) {
                 ensureBuffer();
-                nextBuffer.skip(nextBuffer.available());
+                if (!finished) {
+                    // we may have finished in ensureBuffer()
+                    nextBuffer.skip(nextBuffer.available());
+                }
             }
         }
     }

--- a/webserver/tests/webserver/pom.xml
+++ b/webserver/tests/webserver/pom.xml
@@ -34,6 +34,11 @@
             <artifactId>helidon-webserver</artifactId>
         </dependency>
         <dependency>
+            <groupId>io.helidon.http.media</groupId>
+            <artifactId>helidon-http-media-multipart</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
             <groupId>io.helidon.webserver.testing.junit5</groupId>
             <artifactId>helidon-webserver-testing-junit5</artifactId>
             <scope>test</scope>

--- a/webserver/tests/webserver/src/test/java/io/helidon/webserver/tests/MultipartTest.java
+++ b/webserver/tests/webserver/src/test/java/io/helidon/webserver/tests/MultipartTest.java
@@ -1,0 +1,149 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.webserver.tests;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import io.helidon.common.media.type.MediaTypes;
+import io.helidon.http.Status;
+import io.helidon.http.media.multipart.MultiPart;
+import io.helidon.http.media.multipart.ReadablePart;
+import io.helidon.http.media.multipart.WriteableMultiPart;
+import io.helidon.http.media.multipart.WriteablePart;
+import io.helidon.webclient.http1.Http1Client;
+import io.helidon.webclient.http1.Http1ClientResponse;
+import io.helidon.webserver.http.HttpRules;
+import io.helidon.webserver.http.ServerRequest;
+import io.helidon.webserver.http.ServerResponse;
+import io.helidon.webserver.testing.junit5.ServerTest;
+import io.helidon.webserver.testing.junit5.SetUpRoute;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.hasItems;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@ServerTest
+class MultipartTest {
+    private static final List<String> NAMES = new ArrayList<>();
+    private final Http1Client client;
+
+    MultipartTest(Http1Client client) {
+        this.client = client;
+    }
+
+    @SetUpRoute
+    static void routing(HttpRules rules) {
+        rules.post("/", MultipartTest::multipartRoute);
+    }
+
+    @Test
+    void testMultipleReadParts() throws IOException {
+        try (Http1ClientResponse response = client.post("/")
+                .queryParam("read", "true")
+                .submit(WriteableMultiPart.builder()
+                                .addPart(writeablePart("field_1", "a"))
+                                .addPart(writeablePart("field_2", "b"))
+                                .build())) {
+
+            assertThat(response.status(), is(Status.NO_CONTENT_204));
+            assertThat(NAMES, hasItems("field_1", "field_2"));
+        }
+    }
+
+    @Test
+    void testMultipleSkipReadParts() throws IOException {
+        try (Http1ClientResponse response = client.post("/")
+                .queryParam("read", "false")
+                .submit(WriteableMultiPart.builder()
+                                .addPart(writeablePart("field_1", "a"))
+                                .addPart(writeablePart("field_2", "b"))
+                                .build())) {
+
+            assertThat(response.status(), is(Status.NO_CONTENT_204));
+            assertThat(NAMES, hasItems("field_1", "field_2"));
+        }
+    }
+
+    @Test
+    void testMultipleReadPartsNoLength() throws IOException {
+        try (Http1ClientResponse response = client.post("/")
+                .queryParam("read", "true")
+                .submit(WriteableMultiPart.builder()
+                                .addPart(writeablePartNoLength("field_1", "a"))
+                                .addPart(writeablePartNoLength("field_2", "b"))
+                                .build())) {
+
+            assertThat(response.status(), is(Status.NO_CONTENT_204));
+            assertThat(NAMES, hasItems("field_1", "field_2"));
+        }
+    }
+
+    @Test
+    void testMultipleSkipReadPartsNoLength() throws IOException {
+        try (Http1ClientResponse response = client.post("/")
+                .queryParam("read", "false")
+                .submit(WriteableMultiPart.builder()
+                                .addPart(writeablePartNoLength("field_1", "a"))
+                                .addPart(writeablePartNoLength("field_2", "b"))
+                                .build())) {
+
+            assertThat(response.status(), is(Status.NO_CONTENT_204));
+            assertThat(NAMES, hasItems("field_1", "field_2"));
+        }
+    }
+
+    @BeforeEach
+    void clearNames() {
+        NAMES.clear();
+    }
+
+    private static void multipartRoute(ServerRequest req, ServerResponse res) {
+        boolean readIt = req.query().first("read").asBoolean().orElse(true);
+        MultiPart multiPart = req.content().as(MultiPart.class);
+
+        while (multiPart.hasNext()) {
+            ReadablePart next = multiPart.next();
+            NAMES.add(next.name());
+            if (readIt) {
+                next.consume();
+            }
+        }
+
+        res.status(Status.NO_CONTENT_204).send();
+    }
+
+    private WriteablePart writeablePart(String partName, String content) throws IOException {
+        return WriteablePart.builder(partName)
+                .content(content.getBytes(StandardCharsets.UTF_8))
+                .contentType(MediaTypes.MULTIPART_FORM_DATA)
+                .build();
+    }
+
+    private WriteablePart writeablePartNoLength(String partName, String content) throws IOException {
+        return WriteablePart.builder(partName)
+                .contentType(MediaTypes.MULTIPART_FORM_DATA)
+                .inputStream(() -> new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8)))
+                .build();
+    }
+}

